src/jit/codegenarmarch.cpp

```diff
@@ -859,6 +859,154 @@ void CodeGen::genCodeForShift(GenTreePtr tree)
     genProduceReg(tree);
 }
 
+// Generate code for CpObj nodes wich copy structs that have interleaved
+// GC pointers.
+// For this case we'll generate a sequence of loads/stores in the case of struct
+// slots that don't contain GC pointers.  The generated code will look like:
+// ldr tempReg, [R13, #8]
+// str tempReg, [R14, #8]
+//
+// In the case of a GC-Pointer we'll call the ByRef write barrier helper
+// who happens to use the same registers as the previous call to maintain
+// the same register requirements and register killsets:
+// bl CORINFO_HELP_ASSIGN_BYREF
+//
+// So finally an example would look like this:
+// ldr tempReg, [R13, #8]
+// str tempReg, [R14, #8]
+// bl CORINFO_HELP_ASSIGN_BYREF
+// ldr tempReg, [R13, #8]
+// str tempReg, [R14, #8]
+// bl CORINFO_HELP_ASSIGN_BYREF
+// ldr tempReg, [R13, #8]
+// str tempReg, [R14, #8]
+void CodeGen::genCodeForCpObj(GenTreeObj* cpObjNode)
+{
+    GenTreePtr dstAddr       = cpObjNode->Addr();
+    GenTreePtr source        = cpObjNode->Data();
+    var_types  srcAddrType   = TYP_BYREF;
+    bool       sourceIsLocal = false;
+    regNumber  dstReg        = REG_NA;
+    regNumber  srcReg        = REG_NA;
+
+    assert(source->isContained());
+    if (source->gtOper == GT_IND)
+    {
+        GenTree* srcAddr = source->gtGetOp1();
+        assert(!srcAddr->isContained());
+        srcAddrType = srcAddr->TypeGet();
+    }
+    else
+    {
+        noway_assert(source->IsLocal());
+        sourceIsLocal = true;
+    }
+
+    bool dstOnStack = dstAddr->OperIsLocalAddr();
+
+#ifdef DEBUG
+    assert(!dstAddr->isContained());
+
+    // This GenTree node has data about GC pointers, this means we're dealing
+    // with CpObj.
+    assert(cpObjNode->gtGcPtrCount > 0);
+#endif // DEBUG
+
+#if defined(_TARGET_ARM_)
+    dstReg = REG_ARG_0;
+    srcReg = REG_ARG_1;
+#elif defined(_TARGET_ARM64_)
```

Comment:
> You mean, is it better to implement ASM WRITE BARRIER HELPER first and add that define?

No. Simply add the defines for ARM, not under any kind of `#ifdef`. For example:
```
#define REG_WRITE_BARRIER_SRC_BYREF    REG_ARG_1
#define RBM_WRITE_BARRIER_SRC_BYREF    RBM_ARG_1

#define REG_WRITE_BARRIER_DST_BYREF    REG_ARG_0
#define RBM_WRITE_BARRIER_DST_BYREF    RBM_ARG_0
```

Whether the write barrier helper is implemented with a special ABI or not doesn't matter: we still set up dst and src registers. In the ARM case, they just happen to be in the normal argument registers.